### ISSUE TITLE
feat: add CJS named export annotations for Transform API

### ISSUE THREAD

**Issue by privatenumber** [evanw/esbuild#2248]
## Request

Node.js uses [cjs-module-lexer](https://github.com/nodejs/cjs-module-lexer) to [parse ComonJS named exports](https://github.com/nodejs/node/blob/acc11c98735b15cc3cda6426fb978c620f91522b/lib/internal/modules/esm/translators.js#L218) to interop with ESM imports.

Since esbuild compiles ESM exports to CJS using getters on `module.exports`, it is currently not parsable.

I would like to request a minor syntax hint to help the Node.js detect named exports in the Transform API.

#### ESM Input
```ts
// ts.ts

export const a = 1;
```

#### CJS Output
```js
// ts.cjs

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ts_exports = {};
__export(ts_exports, {
  a: () => a
});
module.exports = __toCommonJS(ts_exports);
const a = 1;
```

#### Importing the output
The named export `a` is not detected by Node.js, and is grouped together behind the `default` export.

```js
// index.mjs

const imported = await import('./ts.cjs');
console.log(imported);
// => [Module: null prototype] { default: { a: [Getter] } }
```

## Proposal
Add a purely syntactical hint to help the lexer detect the named exports.

In this example, I add `exports.<export-name> = void 0` at the end of the transformed code. This doesn't do anything because `module.exports` is overwritten to another object prior.

```ts
// ts.cjs

var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ts_exports = {};
__export(ts_exports, {
  a: () => a
});
module.exports = __toCommonJS(ts_exports);
const a = 1;
exports.a = void 0; // Hint for lexer
```

#### Importing the output
The named export is successfully detected by Node.js:

```js
// index.mjs

const imported = await import('./ts.cjs');
console.log(imported);
// => [Module: null prototype] { a: 1, default: { a: [Getter] } }
```

## Context
We're developing a robust esbuild enhanced Node.js runtime called [tsx](https://github.com/esbuild-kit/tsx) using the transform API. And there is an edge-case for `import()`ing an CJS module (originally ESM, transpiled by esbuild on-demand) from a pure ESM file, but the named exports are consolidated in a `default` export.

Currently, I'm considering using some regex to detect the export getters and injecting the hints at the bottom of the transformed code. However, this could be slow or frail and is not an ideal long-term solution.

I believe adding these hints in esbuild would benefit all use-cases that compile ESM/TS to CJS, and also other tools that are targeted for Node.

## Alternative ideas
I would also appreciate a list of exports from the transform API, similar to what [`es-module-lexer`](https://github.com/guybedford/es-module-lexer) does. That way, the hints can be added on our end reliably.

Using `es-module-lexer` to detect exports is how we [detect whether a module is ESM](https://github.com/esbuild-kit/esm-loader/blob/0b61f19ab7b48f2ed15b9992ff224d5538ed7027/src/loaders.ts#L186-L188) so it would also help remove double-parsing.

**Comment by hyrious** [comment]
There was already a hint made for nodejs when you enabled `platform: "node"` (only available in _build_ mode): [(see it online)](https://hyrious.me/esbuild-repl/?version=0.14.39&mode=build&modules=%5B%5B%22main.js%22%2C%22export+let+a+%3D+1%22%2C1%5D%5D&buildOptions=%7B%22bundle%22%3Atrue%2C%22platform%22%3A%22node%22%7D)

https://github.com/evanw/esbuild/blob/a599d45c77f91ee4c4238ce679abc832aed579f0/internal/bundler/linker.go#L4177

**Comment by privatenumber** [comment]
Thanks @hyrious. That's exactly what I was looking for! Glad it's already been implemented.

However, this only seems available for the Build API and we're using the Transform API.

**Comment by privatenumber** [comment]
I investigated further and I think I may have misinterpreted the problem. Closing in favor of https://github.com/evanw/esbuild/issues/2253

---

For reference, I was misinterpreting Node.js behavior when a CJS file uses `import()` to import another CJS file (transformed from ESM).

_index.js_ (gets transformed to CJS)
```js
import('./esm.js').then(m => console.log(m));
```

_esm.js_ (gets transformed to CJS)
```js
export const a = 1;
```

In this case, Node.js [reads the file source directly from disk](https://github.com/nodejs/node/blob/bde889bd4ed68d33163897b4175557fbc669b66f/lib/internal/modules/esm/translators.js#L211) so cjs-module-lexer can't detect the CJS exports because it would be parsing ESM. Therefore, adding named export annotations didn't actually make a difference.
